atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java

```diff
@@ -0,0 +1,161 @@
+import com.palantir.common.base.Throwables;
+
+public class PersistentLock {
+    private static final long LOCKS_TIMESTAMP = 0L;
+
+    private static final Logger log = LoggerFactory.getLogger(PersistentLock.class);
+
+    private final KeyValueService keyValueService;
+
+    public PersistentLock(KeyValueService keyValueService) {
+        this.keyValueService = keyValueService;
+
+        createPersistedLocksTableIfNotExists();
+    }
+
+    public <T> T runWithExclusiveLock(
+            Supplier<T> supplier,
+            PersistentLockName lock,
+            String reason) throws PersistentLockIsTakenException {
+        return runWithLockInternal(supplier, lock, reason, true);
+    }
+
+    public <T> T runWithNonExclusiveLock(
+            Supplier<T> supplier,
+            PersistentLockName lock,
+            String reason) throws PersistentLockIsTakenException {
+        return runWithLockInternal(supplier, lock, reason, false);
+    }
+
+    private <T> T runWithLockInternal(
+            Supplier<T> supplier,
+            PersistentLockName lock,
+            String reason,
+            boolean exclusive) throws PersistentLockIsTakenException {
+        LockEntry acquiredLock = acquireLock(lock, reason, exclusive);
+
+        try {
+            return supplier.get();
+        } catch (Exception e) {
+            throw Throwables.throwUncheckedException(e);
+        } finally {
+            releaseLock(acquiredLock);
+        }
+    }
+
+    public LockEntry acquireLock(PersistentLockName lockName, String reason) throws PersistentLockIsTakenException {
+        return acquireLock(lockName, reason, true);
+    }
+
+    public LockEntry acquireLock(
+            PersistentLockName lockName, String reason, boolean exclusive) throws PersistentLockIsTakenException {
+        long thisId = ThreadLocalRandom.current().nextLong(Long.MAX_VALUE);
+        LockEntry lockEntry = LockEntry.of(lockName, thisId, reason, exclusive);
+
+        insertLockEntry(lockEntry);
+
+        return verifyLockWasAcquired(lockEntry);
+    }
+
+    public void releaseLock(LockEntry lock) {
+        log.debug("Releasing persistent lock " + lock);
+        keyValueService.delete(AtlasDbConstants.PERSISTED_LOCKS_TABLE, lock.deletionMapWithTimestamp(LOCKS_TIMESTAMP));
+    }
+
+    private LockEntry verifyLockWasAcquired(LockEntry desiredLock) throws PersistentLockIsTakenException {
+        List<LockEntry> relevantLocks = allRelevantLockEntries(desiredLock.lockName());
+        Preconditions.checkState(relevantLocks.contains(desiredLock), "Lock was not properly inserted");
+
+        List<LockEntry> conflictingLocks = removeSingleLock(desiredLock, relevantLocks);
+
+        if (!desiredLock.exclusive()) {
+            conflictingLocks = retainExclusiveLocks(conflictingLocks);
+        }
+        return verifyLockDoesNotConflict(desiredLock, conflictingLocks);
+    }
+
+    private List<LockEntry> retainExclusiveLocks(List<LockEntry> conflictingLocks) {
```

Comment:
conflicting LockEntrys might hashCode / equals the same?
or maybe we care about ordering here for who wins a conflict